File: src/engsoft/jogo/patos/Pato.java
package engsoft.jogo.patos;

public abstract class Pato {
    
    // Corrigido: agora são dois atributos distintos e com nomes consistentes
    protected Padrao_Voaveis comportamento_pato_voo;
    protected Padrao_Pular comportamento_pato_pulo;

    // Método abstrato, implementado nas subclasses
    abstract String mostrar();

    // Método comum a todos os patos
    public String nadar() {
        return "Pato Nadando.";
    }

    // Define os comportamentos dinamicamente
    public void setComportamento(Padrao_Voaveis padrao_voo, Padrao_Pular padrao_pulo) {
        this.comportamento_pato_voo = padrao_voo;
        this.comportamento_pato_pulo = padrao_pulo;
    }

    // Executa os comportamentos atuais e retorna as ações
    public String comportamento_pato() {
        String voo = (comportamento_pato_voo != null) ? comportamento_pato_voo.voar() : "Não sabe voar.";
        String pulo = (comportamento_pato_pulo != null) ? comportamento_pato_pulo.pular() : "Não sabe pular.";
        return voo + " " + pulo;
    }
}

File: Pato.ts
import { Padrao_Voaveis } from "./Padrao_Voaveis";
import { Padrao_Pular } from "./Padrao_Pular";

export abstract class Pato {
    protected comportamento_pato_voo: Padrao_Voaveis | null = null;
    protected comportamento_pato_pulo: Padrao_Pular | null = null;

    // Método abstrato
    abstract mostrar(): string;

    // Método comum
    nadar(): string {
        return "Pato Nadando.";
    }

    // Define os comportamentos dinamicamente
    setComportamento(padrao_voo: Padrao_Voaveis, padrao_pulo: Padrao_Pular): void {
        this.comportamento_pato_voo = padrao_voo;
        this.comportamento_pato_pulo = padrao_pulo;
    }

    // Executa os comportamentos atuais
    comportamento_pato(): string {
        const voo = this.comportamento_pato_voo ? this.comportamento_pato_voo.voar() : "Não sabe voar.";
        const pulo = this.comportamento_pato_pulo ? this.comportamento_pato_pulo.pular() : "Não sabe pular.";
        return `${voo} ${pulo}`;
    }
}
